bin/build-calypso-app.mjs

```diff
@@ -0,0 +1,121 @@
+import chokidar from 'chokidar';
+import runAll from 'npm-run-all';
+import yargs from 'yargs';
+import { hideBin } from 'yargs/helpers';
+
+// Default NODE_ENV is development unless manually set to production.
+if ( ! process.env.NODE_ENV ) {
+	process.env.NODE_ENV = 'development';
+}
+if ( ! process.env.BROWSERSLIST_ENV ) {
+	process.env.BROWSERSLIST_ENV = 'wpcom';
+}
+
+const { argv } = yargs( hideBin( process.argv ) ).options( {
+	watch: { type: 'boolean', default: false },
+	sync: { type: 'boolean', default: false, implies: 'remotePath' },
+	remotePath: { type: 'string' },
+	localPath: {
+		type: 'string',
+		required: true,
+		coerce: ( distPath ) => ( distPath ? `${ process.cwd() }/${ distPath }/` : undefined ),
+	},
+	verbose: { type: 'boolean', default: false, alias: 'v' },
+} );
+await runBuilder( argv );
+
+async function runBuilder( { watch, sync, localPath, remotePath, verbose } ) {
+	if ( verbose ) {
+		console.info( `Watch mode: ${ watch ? 'yes' : 'no' }\nSync mode: ${
+			sync ? 'yes' : 'no'
+		}\nLocal path: ${ localPath }\nRemote path: ${ remotePath }
+		` );
+	}
+
+	const runOpts = {
+		stdout: process.stdout,
+		stderr: process.stderr,
+		printLabel: true,
+	};
+	console.log( 'Cleaning...' );
+	await runAll( 'clean', runOpts );
+
+	console.log( 'Starting webpack...' );
+	// Run build tasks in parallel
+	runAll( [ 'build:*' ], { ...runOpts, parallel: true } )
+		.then( () => {
+			console.log( 'Build completed!' );
+			if ( ! watch && sync ) {
+				// In non-watch + sync mode, we sync only once after the build has finished.
+				syncToRemote( localPath, remotePath );
+			}
+		} )
+		.catch( ( e ) => {
+			console.error( 'The build failed.' );
+			console.error( `Reported build error: ${ e.message }` );
+			showTips( e.results );
+			process.exit( 1 );
+		} );
+
+	// In dev mode, we start watching to sync while the webpack build is happening.
+	if ( watch && sync ) {
+		syncToRemote( localPath, remotePath, true );
+	}
+}
+
+function syncToRemote( localPath, remotePath, watch = false ) {
+	const debouncedSync = debouncer( () => {
+		console.info( 'Performing sync...' );
+		exec( `rsync -ahz --exclude=".*" ${ localPath } wpcom-sandbox:${ remotePath }` );
+	} );
+	if ( watch ) {
+		chokidar.watch( localPath ).on( 'change', debouncedSync );
+	} else {
+		debouncedSync();
+	}
+}
+
+// A debouncer that does the following:
+// - Calls the func immediately if no calls happened in the past 500ms. (Such as
+//   the first time the debounced function is called.)
+// - Calls func again only after further calls have stopped for 500ms. (Does not
+//   call func if no calls happened after the first time.)
+function debouncer( func, delayMs = 500 ) {
+	let timeout = null;
+	let calledWithTimeout = false;
+	return () => {
+		if ( ! timeout ) {
+			func();
+		} else {
+			calledWithTimeout = true;
+		}
+		clearTimeout( timeout );
+		timeout = setTimeout( () => {
+			if ( calledWithTimeout ) {
+				func();
+				calledWithTimeout = false;
```

Comment:
I think it would still run after the next timeout runs out. The next time the debounced function is called, another timeout will still get set. (it won't get called on the leading edge though!)

that said, we don't need this logic if we stop calling on the leading edge.

I originally added this since I was thinking about https://github.com/Automattic/wp-calypso/issues/48830. But I think those are two different problems